### PR TITLE
Don't lint autolinks in `doc_markdown`

### LINKED ISSUE #1832: doc_markdown should ignore plain URLs
The `doc_markdown` rule is expected to search the *plain text* of a Markdown content, and should exclude non-plain-text parts, like code blocks and *URLs*.

Plain URLs (not labeled links or URLs in brackets) are core components of markdown and should be respected as URLs and not plain text.

From the perspective of what's going on in practice: when implementing a spec in code, it's very common to put a plain URL link to the spec, as a point of reference. These URLs are most of the time very well self-descriptive and there's no need for extra work to *label* them.

Example: all the URLs like `http://www.unicode.org/reports/tr9/...` here: https://github.com/servo/unicode-bidi/blob/master/src/explicit.rs

Based on these, clippy should no warn about any formatting issues about pain URLs, specially the backtick requirement.

Also reported back in Feb: https://github.com/Manishearth/rust-clippy/issues/1558

### PR DESCRIPTION
**Pull request by mcarton**
Close #1832.

### CODE REVIEW

File: clippy_tests/examples/doc.stderr [hunk 1]
```diff
@@ -206,6 +206,14 @@ error: you should put `be_sure_we_got_to_the_end_of_it` between ticks in the doc
     |
     = note: `-D doc-markdown` implied by `-D warnings`
 
+error: you should put `http://www.unicode.org/reports/tr9/#Reordering_Resolved_Levels` between ticks in the documentation
```
Comment (oli-obk) [inline clippy_tests/examples/doc.stderr]: I think we should add some simple heuristic to suggest surrounding it with `<>`. Something like "anything that begins with `{ascii}*://` or `www.`"

Comment (killercup) [inline clippy_tests/examples/doc.stderr]: I'd suggest

```rust
extern crate url;
use url::Url;

let is_url = Url::parse(&content).is_ok();

format!("You should put {wrap1} around `{content}`  to mark it as {thing1},\
         or {wrap2} to mark it as {thing2}",
    content = content,
    wrap1 = if is_url { "<...>" } else { "`...`" }, thing1 = if is_url { "URL" } else { "code" },
    wrap2 = if !is_url { "<...>" } else { "`...`" }, thing2 = if !is_url { "URL" } else { "code" },
)
```

Comment (oli-obk) [inline clippy_tests/examples/doc.stderr]: Now I want a lint that detects code which applies a heuristic to text to find urls and suggests to use the `url` crate instead!

Comment (mcarton) [inline clippy_tests/examples/doc.stderr]: Done. I meant for the first part, for the "lint that detects code which applies a heuristic to text to find urls and suggests to use the `url` crate instead" we'll see in another PR 😆

File: clippy_lints/src/doc.rs [hunk 2]
```diff
@@ -258,6 +259,18 @@ fn check_word(cx: &EarlyContext, word: &str, span: Span) {
         s != "_" && !s.contains("\\_") && s.contains('_')
     }
 
+    if let Ok(url) = Url::parse(word) {
```
Comment (Manishearth) [inline clippy_lints/src/doc.rs]: I feel like we should first check for http/https or something to avoid hitting the URL parser too much

Comment (llogiq) [inline clippy_lints/src/doc.rs]: We may want to have a configurable list of protocols, prepopulated with `["file", "http", "https", "ftp", "ssh", "mailto"]`.

Comment (oli-obk) [inline clippy_lints/src/doc.rs]: I think it's fine as it is... we have many things in clippy that we should benchmark, we shouldn't optimize before actually doing that. I'd wager the URL parser isn't much worse in perf than us adding a quick check

**Comment by oli-obk** [comment]
needs a rebase

**Comment by oli-obk** [comment]
ping @mcarton

**Comment by lucab** [comment]
Bump on this, any chance to get a rebase?

**Comment by mcarton** [comment]
Rebased, sorry it took so long, I had completely forgotten about this

**Pull request merged**